
<h2>Problem</h2>
<p>You are using the <a href="/docs/release-notes/agent-release-notes/python-release-notes">latest Python agent</a>, and you see a log entry with a warning message such as:</p>
<pre><code>Attempt to activate application in a process different to where the agent harvest thread was started.
</code></pre>
<p>OR:</p>
<pre><code>Attempt to reactivate application or record transactions in a process different to where the agent was already registered.
</code></pre>
<p>The application shows as reporting in the New Relic UI, but no data is reported to New Relic.</p>
<h2>Solution</h2>
<p>To solve this problem with your app:</p>
<ol>
  <li>
    <p>Move all calls to <a href="/docs/agents/python-agent/python-agent-api/register_application"><code>newrelic.agent.register_application</code></a> or <a href="/docs/agents/python-agent/python-agent-api/application"><code>newrelic.agent.application</code></a> inside of functions inside of a <code>__name__ == __main__</code> check.</p>
  </li>
  <li>
    <p>If you're unsure where the calls to <a href="/docs/agents/python-agent/python-agent-api/register_application"><code>newrelic.agent.register_application</code></a> or <code>newrelic.agent.application</code> are occurring, use the <a href="/docs/agents/python-agent/troubleshooting/python-agent-logging">agent debug logs</a> to search for an entry containing:</p>
    <pre><code>newrelic.core.agent DEBUG - Application was activated from:
</code></pre>
  </li>
  <li>
    <p>Use the entry's traceback of the call that activated the agent. Refer to the following frames, which are considered normal:</p>
    <pre><code>File "newrelic/api/transaction.py", line x, in __init__
    application.activate()
</code></pre>
    <pre><code>File "newrelic/hooks/application_celery.py", line x, in process_initializer
    application_instance().activate()
</code></pre>
  </li>
  <li>
    <p>If the activation is occurring from a different location, correct this issue by following this example:</p>
    <p><strong>Before:</strong></p>
    <pre><code>import newrelic.agent

# This will cause the agent to activate whenever custom_event is imported
app = newrelic.agent.application()

def custom_event():
    newrelic.agent.record_custom_event('CustomEvent', {}, application=app)
</code></pre>
    <p><strong>After</strong>:</p>
    <pre><code>import newrelic.agent

def custom_event():
    app = newrelic.agent.application()
    newrelic.agent.record_custom_event('CustomEvent', {}, application=app)
</code></pre>
  </li>
</ol>
<p>For other tips when no data appears, see the <a href="/docs/agents/python-agent/troubleshooting/no-data-appears-python">Python troubleshooting documentation</a>.</p>
<h2>Cause</h2>
<p>This is usually caused by a call to <a href="/docs/agents/python-agent/python-agent-api/register_application"><code>newrelic.agent.register_application</code></a> or <a href="/docs/agents/python-agent/python-agent-api/application"><code>newrelic.agent.application</code></a> occurring at import time.</p>
<p><strong>Example</strong>:</p>
<pre><code>import newrelic.agent

# This will cause the agent to activate whenever custom_event is imported
app = newrelic.agent.application()

def custom_event():
    newrelic.agent.record_custom_event('CustomEvent', {}, application=app)
</code></pre>
<p>Code typically is loaded before a fork occurs in frameworks such as Celery or multiprocessing. As a result, the code is imported into theÂ parent process instead of the child, launching the agent. Since the agent is launched in the parent process, no data is collected in the child process.</p>
